M2-4835: Linting rules for `mindlogger-admin`

sultanofcardio opened the pull request:
resolves: [M2-4835](https://mindlogger.atlassian.net/browse/M2-4835)

This PR creates a GitHub Action to run linting tasks on pull requests submitted to the repository. As part of the ticket, I took the opportunity to update the ESlint rules on the project. These updates have raised a number of linting issues, but **won't block** the merging of pull requests. While I've addressed some of the smaller items (and any error ones), I expect that these will be addressed in the normal course of feature development so that we can eventually require linting checks to pass in order to merge PRs.

Here's a summary of what changed:

## Conflicting ESLint configurations

There are currently two ESLint configurations in the project:
- An `eslintConfig` property in the `package.json` file
- A `.eslintrc` file

[ESLint documentation](https://eslint.org/docs/latest/use/configure/configuration-files) says that only one of these is read, and the `.eslintrc` takes precedence. Thus, I've removed the ignored configuration since it wasn't being used.

## Prettier Code Formatting

Prettier has been configured in the project as the code formatter of choice, but it is currently conflicting with a number of code formatting ESLint rules. The prettier ESLint plugin disables these rules if it is placed as an inherited configuration at the end of the `extends` array, so I have made this change. The prettier ESLint plugin also enables Prettier as an ESLint rule, so that code formatting issues become errors. This should be okay, since Prettier fixes them for you.

In addition, I've disabled/removed some other code formatting related rules that aren't explicitly handled by the Prettier plugin. These include:
- `padding-line-between-statements`
- `no-extra-semi`
- `jsx-quotes` - I've moved this one to the Prettier config with `jsxSingleQuote`

## Opinionated changes

There are a number of other changes I've made that I think will allow for improved code quality. These are:
- Turning on the `react-hooks/exhaustive-deps` rule
- Adding the `unused-imports` ESLint plugin
- Turning on the `@typescript-eslint/no-non-null-assertion` rule

[M2-4835]: https://mindlogger.atlassian.net/browse/M2-4835?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sultanofcardio committed:
M2-4835: Remove unused eslintConfig

sultanofcardio committed:
M2-4835: Add linting scripts

sultanofcardio committed:
M2-4835: Remove unnecessary unused variable

sultanofcardio committed:
M2-4835: Remove unused import

sultanofcardio committed:
M2-4835: Convert enum to const object

This is to satisfy the `@typescript-eslint/no-duplicate-enum-values` ESLint rule

sultanofcardio committed:
M2-4835: Replace "prettier" with "plugin:prettier/recommended"

This plugin adds prettier as an ESLint rule, so that files not properly formatted according to the prettier config will get flagged as ESLint errors.

I also moved the configuration to the end of the array, to allow it to turn off preceding rules (as it does).

Finally, I removed the remaining rules that conflict with those from `eslint-config-prettier`, namely "quotes", "no-trailing-spaces", and "arrow-parens". I also moved "jsx-quotes" to the prettier config

sultanofcardio committed:
M2-4835: Disable upcoming deprecated rules

These rules are deprecated in ESLint v8.53.0

sultanofcardio committed:
M2-4835: Turn "react-hooks/exhaustive-deps" to warn

sultanofcardio committed:
M2-4835: Add unused-imports plugin with corresponding rules

sultanofcardio committed:
M2-4835: Turn "@typescript-eslint/no-non-null-assertion" to warn

sultanofcardio committed:
M2-4835: Add explicit dependencies

The project depends on eslint-plugin-react and eslint-plugin-import, but they are not in the package.json, so I added them

sultanofcardio committed:
M2-4835: Update CI workflow

sultanofcardio committed:
M2-4835: Fix more prettier issues

sultanofcardio committed:
M2-4835: Fail the Lint job if there are linting errors

The linting job creates check runs (ESLint, Prettier) which detail their respective outcomes, but the job itself does not fail if there are linting errors. This changes that

sultanofcardio committed:
M2-4835: Unify node versions among jobs and add dependency caching

sultanofcardio committed:
M2-4835: Bring Prettier in line with mindlogger-web-refactor

sultanofcardio committed:
M2-4835: Bring in more ESLint changes from mindlogger-web-refactor

sultanofcardio committed:
M2-4835: Disable no-underscore-dangle

alex-i-e commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,55 @@
+on: [pull_request]
+
+env:
+  NODE_VERSION: 20.11.0
```
it should be mentioned that all admin panel team works with nodejs v.18 . Maybe it must be specified in package.json and here as well to reduce possible errors relating to different versions of nodejs

sultanofcardio replied:
Thanks for clearing that up. I'll revise the node version

IhorNesterenko99 commented on .eslintrc:
```diff
@@ -1,51 +1,51 @@
+      {
+        "groups": [["external", "builtin"], "internal", ["sibling", "parent"], "index"],
+        "pathGroups": [
+          {
```
I believe that the empty line between the React import and other imports from libraries, as well as the alphabetized ordering of imports, may be considered excessive. Would you be able to provide some insight into the reasoning behind these choices?

alex-i-e replied:
It leads to changes almost all the files and leads to investigate more time to figure out who did the initial changes and why.

sultanofcardio replied:
I took these rules from the `mindlogger-app-refactor` repo.

https://github.com/ChildMindInstitute/mindlogger-app-refactor/blob/a5c8f472cfefe2c5db6ccd0d02efafe73381d8d8/.eslintrc.js#L42-L46

Part of this effort is to bring the code styles in line with each other, to lower the cognitive load of switching between repos. I don't have a preference either way because the tools will auto-format the files which means we won't have to think about it.

sultanofcardio replied:
Or maybe it was the `mindlogger-web-refactor`, I don't remember but they both use it

sultanofcardio replied:
As it turns out, alphabetized imports break the test files because they currently depend on import order, so I've reverted this for now

alex-i-e replied:
Thank you, and could you please revert adding empty line after `react` imports (by keeping it in group of `external` imports on the top of the files)?

IhorNesterenko99 commented on .eslintrc:
```diff
@@ -1,51 +1,51 @@
     "import/no-cycle": "error",
     "react-hooks/rules-of-hooks": "error",
-    "react-hooks/exhaustive-deps": "off",
+    "react-hooks/exhaustive-deps": "warn",
```
In numerous cases, not all arguments are necessary to be included in the dependencies of the useEffect hook. Implementing this rule might result in an abundance of unexpected warnings, which could be perceived as excessive.

alex-i-e replied:
It will lead to add exceptional cases when we don't need to add dependencies as deps hooks because of existed business logic. For example, we need to add some useEffect on mount / unmount and to highlight it we don't need to add dependencies there.

sultanofcardio replied:
I'm not sure I understand the comment. The cases where it is necessary to exclude dependencies from the array should be the exception, and in those cases the rule can be disabled. Not having warnings at all increases the likelihood that we miss cases where omitting dependencies isn't the intention.

I don't see the high number of warning as a problem, because we won't block PRs for warnings. It is simply a signal to the developer (and reviewer) that you should pay attention here

farmerpaul replied:
Missing dependencies can be incredibly annoying to debug, and I agree with @sultanofcardio that this rule is essential, even if enabled just as a warning for now (though if it were up to me, I'd want it to issue an error).

If you're certain you want to exclude a variable from the dependencies, rather than disable this rule entirely via a `// es-disable-…` comment, you can opt to do a more surgical approach by placing that variable in a ref, for example:

```tsx
export const MyComponent: React.FC<MyComponentProps> = ({ sideEffect }) => {
  const [count, setCount] = useState(0);
  const sideEffectRef = useRef(sideEffect); // 👈 Place unchanging variable in a ref
  
  const handleClick = useCallback(() => {
    setCount(count + 1);
    sideEffectRef.current();
  }, [count]); // 👈 Now we can preserve the exhaustive deps rule here
  
  return <h1 onClick={handleClick}>It Worked: {count)</h1>;
}；
```

If running an effect on mount/unmount is common, one way to reduce instances of `// es-disable-…` is to define a custom hook that is explicit in its purpose, for example:

```tsx
export const useOnceEffect = (callback: React.EffectCallback) => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(callback, []);
};
```

IhorNesterenko99 commented on .prettierrc:
```diff
@@ -1,6 +1,13 @@
 {
-  "semi": true,
-  "trailingComma": "all",
+  "arrowParens": "avoid",
```
My recommendation is to utilize "arrowParens": "always".
Here's the rationale from the Prettier documentation:
"At first glance, avoiding parentheses may look like a better choice because of less visual noise. However, when Prettier removes parentheses, it becomes harder to add type annotations, extra arguments or default values as well as making other changes. Consistent use of parentheses provides a better developer experience when editing real codebases, which justifies the default value for the option."

sultanofcardio replied:
You're right. I must've accidentally overwritten this one with a value from the app repo. I think keeping parentheses is the right move. Thanks for flagging this

IhorNesterenko99 commented on .eslintrc:
```diff
@@ -72,26 +72,20 @@
+    "@typescript-eslint/no-non-null-assertion": "warn",
     "@typescript-eslint/no-explicit-any": "warn",
     "@typescript-eslint/no-duplicate-enum-values": "warn",
-    "padding-line-between-statements": [
```
To enhance code readability, it's commonly suggested to include a blank line between your code and a return statement. Implementing this practice can contribute to better code understanding.

sultanofcardio replied:
I removed this rule because ESLint deprecated and removed it to a separate plugin in the next version. The reason they removed it is because it has to do with opinionated code formatting, for which we are using prettier.

My personal preference is for keeping this so I'll revert for now and see what others think

IhorNesterenko99 commented on .prettierrc:
```diff
@@ -1,6 +1,13 @@
-  "semi": true,
-  "trailingComma": "all",
+  "arrowParens": "avoid",
+  "bracketSameLine": true,
```
Could we consider keeping "bracketSameLine" as false to maintain better code readability?

sultanofcardio replied:
Sure

alex-i-e commented on .prettierrc:
```diff
@@ -1,6 +1,13 @@
+  "jsxSingleQuote": false,
+  "trailingComma": "all",
+  "semi": true,
+  "printWidth": 120,
```
Could you explain why did you extend width of line? It leads to changes almost all the files and leads to investigate more time to figure out who did the initial changes.

sultanofcardio replied:
Yes I did it to keep consistency with the backend and the web app

alex-i-e replied:
It's ok, but could you please revert printWidth?

alex-i-e commented on .prettierrc:
```diff
@@ -1,6 +1,13 @@
+  "printWidth": 120,
+  "tabWidth": 2,
+  "useTabs": false,
+  "endOfLine": "lf"
```
`endOfLine` uses `lf` by default , it seems no need to add this rule as well as `tabWidth`, `useTabs`, `semi`, `trailingComma`

sultanofcardio replied:
Even though they're the default, they serve to indicate to the developer what's being used without needing to check the documentation. The web app also specifies these so it keep them consistent for folks working between repositories

sultanofcardio committed:
M2-4835: Change node version to 18

sultanofcardio committed:
M2-4835: Change `arrowParens` to always

sultanofcardio committed:
M2-4835: Re-enable padding-line-between-statements

sultanofcardio committed:
M2-4835: Change backetSameLine to false

IhorNesterenko99 commented:
I've noticed that the unit tests are failing after the changes made in the pull request. Could you please take a look and correct any issues with the unit tests?

Additionally, there are some conflicts that need to be resolved before this pull request can be merged. Could you please review and resolve the conflicts accordingly?

IhorNesterenko99 commented on .husky/prepare-commit-msg:
I just wanted to bring to your attention that we have some common rules for committing code changes in our project. These rules include including the name of the associated ticket in Jira and adhering to specific branch naming conventions:
- Type: feature, fix, tests, release
- Ticket name in Jira

These guidelines help maintain consistency and clarity across our codebase, reducing any discrepancies between developers when creating branches and writing commit messages.

Could you please share your reasoning behind the decision to delete this code?

sultanofcardio replied:
This is a mistake. I deleted these because the pre-commit hook makes rebasing very tedious, but I didn't intend to push this commit to GitHub. I will remove this commit 👍

sultanofcardio committed:
M2-4835: Fix ESLint and prettier issues after rebase

sultanofcardio commented:
> I've noticed that the unit tests are failing after the changes made in the pull request. Could you please take a look and correct any issues with the unit tests?
>
> Additionally, there are some conflicts that need to be resolved before this pull request can be merged. Could you please review and resolve the conflicts accordingly?

@IhorNesterenko99 Yes, I will address these. I initially believed that the tests were failing because of [this commit](https://github.com/ChildMindInstitute/mindlogger-admin/commit/db0ae5d230d3c887d9b32b2d5aebe6e7c60073ec), but after rebasing again I realised they are still failing.

sultanofcardio committed:
M2-4835: Rollback alphabetical import ordering

It seems that some of the test files depend on importing mocks and other things in a specific order, so I'll revert this for now

sultanofcardio committed:
M2-4835: Revert print width to 100

IhorNesterenko99 commented:
I wanted to bring to your attention that our Jenkins continuous integration is currently failing. Here's a screenshot for reference: 
![image](https://github.com/ChildMindInstitute/mindlogger-admin/assets/116339462/534f5cf0-e995-4eb2-a8de-882d264f365e)
It might be a good idea to reach out to our DevOps lead, Raman Siauko, to investigate what might have caused this issue.

sultanofcardio committed:
M2-4835: Format files with Prettier again

sultanofcardio committed:
M2-4835: Revert import order ESLint rule changes

sultanofcardio committed:
M2-4835: Update package-lock.json with node version

sultanofcardio committed:
M2-4835: Remove prettier from ESLint

Prettier apparently recommends no integrating prettier with linters at all. It also seems that this integration no longer works well (or I've misconfigured them) because prettier and ESLint are making conflicting changes. Therefore, I've opted to run them separately

sultanofcardio committed:
M2-4835: Remove unnecessary ESLint arg in lint job

I also included all the files from `.lintstagedrc.json`

sultanofcardio committed:
M2-4835: Include all the files from `.lintstagedrc.json` in prettier lint job

sultanofcardio committed:
M2-4835: Try replacing `wearerequired/lint-action@v2` with raw commands

This action seems to be reporting Prettier issues where there are none, so I'm trying something else

sultanofcardio committed:
M2-4835: Prefix commands with npx

sultanofcardio committed:
M2-4835: Fix prettier lint issues

As it turns out I was accidentally using a global prettier binary (v2.7.1) instead of the one from the project, which has led to a linting headache for the last few hours

sultanofcardio committed:
M2-4835: Replace the `wearerequired/lint-action@v2` job step

alex-i-e reviewed (approved):
Perfect! You've added meaningful and valuable lint rules for ongoing on project, some of them will point on places that should be fixed in future. I appreciate your input and contribution to improving quality!

sultanofcardio committed:
M2-4835: Re-lint after rebase

sultanofcardio committed:
M2-4835: Re-lint again
